tests/shadowcasting_test.cpp

```diff
@@ -407,199 +408,591 @@ static constexpr float O = LIGHT_TRANSPARENCY_SOLID;
+        this->default_value = default_value;
+    }
 
-    int height() const {
+    int depth() const {
         return data.size();
     }
-    int width() const {
+    int height() const {
         if( data.empty() ) {
             return 0;
         }
         return data[0].size();
     }
+    int width() const {
+        if( data.empty() || data[0].empty() ) {
+            return 0;
+        }
+        return data[0][0].size();
+    }
+    tripoint get_max() const {
+        return offset + tripoint( width(), height(), depth() );
+    }
 
-    float get_global( const point &p ) const {
+    float get_global( const tripoint &p ) const {
         if( p.y >= offset.y && p.y < offset.y + height() &&
-            p.x >= offset.x && p.x < offset.x + width() ) {
-            return data[ p.y - offset.y ][ p.x - offset.x ];
+            p.x >= offset.x && p.x < offset.x + width() &&
+            p.z >= offset.z && p.z < offset.z + depth() ) {
+            return data[ p.z - offset.z ][ p.y - offset.y ][ p.x - offset.x ];
         }
         return default_value;
     }
-
-    float get_local( const point &p ) const {
-        return data[ p.y ][ p.x ];
+    bool get_floor( const tripoint &p ) const {
+        if( !floor.empty() &&
+            p.y >= offset.y && p.y < offset.y + height() &&
+            p.x >= offset.x && p.x < offset.x + width() &&
+            p.z >= offset.z && p.z < offset.z + depth() ) {
+            return data[ p.z - offset.z ][ p.y - offset.y ][ p.x - offset.x ];
+        }
+        return default_floor;
     }
 };
 
-static void run_spot_check( const grid_overlay &test_case, const grid_overlay &expected_result )
+static void run_spot_check( const grid_overlay &test_case, const grid_overlay &expected,
+                            bool fov_3d )
 {
-    float seen_squares[ MAPSIZE * SEEY ][ MAPSIZE * SEEX ] = {{ 0 }};
-    float transparency_cache[ MAPSIZE * SEEY ][ MAPSIZE * SEEX ] = {{ 0 }};
-
-    for( int y = 0; y < static_cast<int>( sizeof( transparency_cache ) /
-                                          sizeof( transparency_cache[0] ) ); ++y ) {
-        for( int x = 0; x < static_cast<int>( sizeof( transparency_cache[0] ) /
-                                              sizeof( transparency_cache[0][0] ) ); ++x ) {
-            transparency_cache[ y ][ x ] = test_case.get_global( point( x, y ) );
+    // Reminder to not trigger 2D shadowcasting on 3D use cases.
+    if( !fov_3d ) {
+        REQUIRE( test_case.depth() == 1 );
+    }
+    level_cache *caches[OVERMAP_LAYERS];
+    std::array<float ( * )[MAPSIZE *SEEX][MAPSIZE *SEEY], OVERMAP_LAYERS> seen_squares;
+    std::array<const float ( * )[MAPSIZE *SEEX][MAPSIZE *SEEY], OVERMAP_LAYERS> transparency_cache;
+    std::array<const bool ( * )[MAPSIZE *SEEX][MAPSIZE *SEEY], OVERMAP_LAYERS> floor_cache;
+
+    const int upper_bound = fov_3d ? OVERMAP_LAYERS : 12;
+    for( int z = fov_3d ? 0 : 11; z < upper_bound; ++z ) {
+        caches[z] = new level_cache();
```

Comment:
Can we avoid raw `new`?  Use `unique_ptr`?  I guess it's a bit annoying because you need the array of pointers, so you'd need a separate container for the `unique_ptr`s...

Reply:
With the way I'm mapping between the contents of the level cache and the relatively large type definitions and establishing aliases for them, this is *really* touchy to adjust, so I'd rather not open that can of worms for an isolated invocation of new in some test code.
I 100% agree that I don't want any avoidable new in the main game code.